kratos/includes/model_part.h

```diff
@@ -357,42 +357,39 @@ class KRATOS_API(KRATOS_CORE) ModelPart final
     void AddNodes(TIteratorType nodes_begin,  TIteratorType nodes_end, IndexType ThisIndex = 0)
     {
         KRATOS_TRY
-        ModelPart::NodesContainerType  aux;
-        ModelPart::NodesContainerType  aux_root; //they may not exist in the root
+        // We need to put the found model part specific nodes into a pointer vector set so
+        // that the insertion ad the current_part->Nodes().insert(aux); doesn't have
+        // to sort it for each parent model part which is redundant.
+        NodesContainerType  aux;
+        std::vector<NodeType::Pointer>  aux_root; //they may not exist in the root
         ModelPart* root_model_part = &this->GetRootModelPart();
 
         for(TIteratorType it = nodes_begin; it!=nodes_end; it++)
         {
-            auto it_found = root_model_part->Nodes().find(it->Id());
+            auto it_found = root_model_part->Nodes().find((*(it.base()))->Id());
             if(it_found == root_model_part->NodesEnd()) //node does not exist in the top model part
             {
                 aux_root.push_back( *(it.base()) ); //node does not exist
-                aux.push_back( *(it.base()) );
+                aux.insert(aux.end(), *(it.base()) );
             }
             else //if it does exist verify it is the same node
             {
-                if(&(*it_found) != &(*it))//check if the pointee coincides
+                if(&(*it_found) != &(*(*(it.base()))))//check if the pointee coincides
                     KRATOS_ERROR << "attempting to add a new node with Id :" << it_found->Id() << ", unfortunately a (different) node with the same Id already exists" << std::endl;
                 else
-                    aux.push_back( *(it.base()) );
+                    aux.insert(aux.end(), *(it.base()) );
```

Comment:
if you use aux unordered you can push_back into it. 
than you can sort, and finally you can insert it ...